File: dotnet/Mcma.Worker/Jobs/WorkerJobHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Mcma.Core;
using Mcma.Data;

namespace Mcma.Worker
{
    public class WorkerJobHelper<T> where T : Job
    {
        public WorkerJobHelper(
            IDbTable<JobAssignment> table,
            ResourceManager resourceManager,
            WorkerRequest request,
            string jobAssignmentId)
        {
            Table = table;
            ResourceManager = resourceManager;
            Request = request;
            
            JobAssignmentId = jobAssignmentId;
        }

        public WorkerRequest Request { get; }

        public ResourceManager ResourceManager { get; }

        public IDbTable<JobAssignment> Table { get; }

        public string JobAssignmentId { get; }

        public JobAssignment JobAssignment { get; private set; }

        public T Job { get; private set; }

        public JobProfile Profile { get; private set; }

        public JobParameterBag JobInput => Job.JobInput;

        public JobParameterBag JobOutput => Job.JobOutput;

        public async Task InitializeAsync()
        {
            JobAssignment = await Table.UpdateJobStatus<JobAssignment>(JobAssignmentId, JobStatus.Running);

            Job = await ResourceManager.ResolveResourceFromFullUrl<T>(JobAssignment.Job);

            Profile = await ResourceManager.ResolveResourceFromFullUrl<JobProfile>(Job.JobProfile);

            Job.JobOutput = new JobParameterBag();
        }

        public void ValidateJob(IEnumerable<string> supportedProfiles)
        {
            var SupportedJobProfiles = supportedProfiles?.ToArray() ?? new string[0];
            if (!SupportedJobProfiles.Contains(Profile.Name, StringComparer.OrdinalIgnoreCase))
                throw new Exception($"Job profile '{Profile.Name}' is not supported");

            if (Profile.InputParameters != null)
            {
                var missingInputParams =
                    Profile.InputParameters.Where(p => !Job.JobInput.HasProperty(p.ParameterName)).Select(p => p.ParameterName).ToList();
                if (missingInputParams.Any())
                    throw new Exception("One or more required input parameters are missing from the job: " + string.Join(", ", missingInputParams));
            }
        }

        public async Task FailAsync(string error)
        {
            await UpdateJobAssignmentStatusAsync(JobStatus.Failed, error);
        }

        public async Task CompleteAsync()
        {
            await UpdateJobAssignmentOutputAsync();
            await UpdateJobAssignmentStatusAsync(JobStatus.Completed);
        }

        public async Task UpdateJobAssignmentAsync(Action<JobAssignment> update, bool sendNotification = false)
        {
            JobAssignment = await Table.GetAsync(JobAssignmentId);
            if (JobAssignment == null)
                throw new Exception("JobAssignment with id '" + JobAssignmentId + "' not found");

            update(JobAssignment);

            JobAssignment.DateModified = DateTime.UtcNow;
            await Table.PutAsync(JobAssignmentId, JobAssignment);
            
            if (sendNotification)
                await SendNotificationAsync();
        }

        public Task UpdateJobAssignmentOutputAsync()
            => UpdateJobAssignmentAsync(ja => ja.JobOutput = JobOutput);

        public Task UpdateJobAssignmentStatusAsync(string status, string statusMessage = null)
            => UpdateJobAssignmentAsync(
                ja =>
                {
                    ja.Status = status;
                    ja.StatusMessage = statusMessage;
                },
                true);

        public async Task SendNotificationAsync()
        {
            if (ResourceManager != null)
                await ResourceManager.SendNotificationAsync(JobAssignment, JobAssignment.NotificationEndpoint);
        }
    }
}

File: nodejs/base/worker/lib/jobs/worker-job-helper.js
const { JobStatus, JobParameterBag, Utils } = require("@mcma/core");

class WorkerJobHelper {
    constructor(jobType, dbTable, resourceManager, request, jobAssignmentId) {
        jobType = Utils.getTypeName(jobType);
        if (!jobType) {
            throw new Error("Worker job helper requires a valid job type to be specified.");
        }

        let jobAssignment;
        let job;
        let profile;
        let matchedProfileName;

        this.getTable = () => dbTable;
        this.getResourceManager = () => resourceManager;
        this.getRequest = () => request;
        this.getJobAssignmentId = () => jobAssignmentId;
        this.getJobAssignment = () => jobAssignment;
        this.getJob = () => job;
        this.getProfile = () => profile;
        this.getMatchedProfileName = () => matchedProfileName;
        this.getJobInput = () => job.jobInput;
        this.getJobOutput = () => job.jobOutput;

        this.initialize = async () => {
            jobAssignment = await this.updateJobAssignmentStatus(JobStatus.RUNNING);

            job = await resourceManager.resolve(jobAssignment.job);

            profile = await resourceManager.resolve(job.jobProfile);

            job.jobOutput = jobAssignment.jobOutput || new JobParameterBag();
        };

        this.validateJob = (supportedProfiles) => {
            if ((job["@type"] || "").toLowerCase() !== jobType.toLowerCase()) {
                throw new Error("Job has type '" + job["@type"] + "', which does not match expected job type '" + jobType + "'.");
            }

            supportedProfiles = supportedProfiles || [];
            matchedProfileName = supportedProfiles.find(x => x && profile.name.toLowerCase() === x.toLowerCase());
            if (!matchedProfileName) {
                throw new Error(`Job profile "${profile.name}" is not supported.`);
            }

            if (profile.inputParameters) {
                const missingInputParams = [];
                const jobInputKeys = job.jobInput && Object.keys(job.jobInput);
                for (const inputParam of profile.inputParameters) {
                    if (!jobInputKeys || !jobInputKeys.includes(inputParam.parameterName)) {
                        missingInputParams.push(inputParam.parameterName);
                    }
                }
                if (missingInputParams.length > 0) {
                    throw new Error("One or more required input parameters are missing from the job: " + missingInputParams.join(", "));
                }
            }
        };

        this.complete = async () => {
            await this.updateJobAssignmentOutput();
            return await this.updateJobAssignmentStatus(JobStatus.COMPLETED);
        };

        this.fail = async (error) => {
            if (!!error && typeof error !== "string") {
                error = JSON.stringify(error);
            }
            return await this.updateJobAssignmentStatus(JobStatus.FAILED, error);
        };

        this.cancel = async (message) => {
            if (!!message && typeof message !== "string") {
                message = JSON.stringify(message);
            }
            return await this.updateJobAssignmentStatus(JobStatus.CANCELED, message);
        };

        this.updateJobAssignmentOutput = async () => {
            return await this.updateJobAssignment(ja => ja.jobOutput = job.jobOutput);
        };

        this.updateJobAssignmentStatus = async (status, statusMessage) => {
            return await this.updateJobAssignment(
                ja => {
                    ja.status = status;
                    ja.statusMessage = statusMessage;
                },
                true);
        };

        this.updateJobAssignment = async (update, sendNotification = false) => {
            if (typeof update !== "function") {
                throw new Error("update must be a function that modifies the JobAssignment.");
            }
            jobAssignment = await dbTable.get(jobAssignmentId);
            if (!jobAssignment) {
                throw new Error("JobAssignment with id '" + jobAssignmentId + "' not found.");
            }

            update(jobAssignment);

            jobAssignment.dateModified = new Date().toISOString();
            await dbTable.put(jobAssignmentId, jobAssignment);

            if (sendNotification) {
                await this.sendNotification();
            }

            return jobAssignment;
        };

        this.sendNotification = async () => {
            if (resourceManager) {
                await resourceManager.sendNotification(jobAssignment);
            }
        };
    }
}

module.exports = {
    WorkerJobHelper
};
